app/services/appointment_scheduler.rb

```diff
@@ -1,46 +1,45 @@
 class AppointmentScheduler
   def initialize(max_schedule_time_ahead:)
     @max_schedule_time_ahead = if max_schedule_time_ahead.is_a?(ActiveSupport::Duration)
-      max_schedule_time_ahead
-    else
-      max_schedule_time_ahead.days
-    end
+                                 max_schedule_time_ahead
+                               else
+                                 max_schedule_time_ahead.days
+                               end
   end
 
   def schedule(raw_start_time:, ubs:, patient:)
     start_time = Time.parse(raw_start_time)
 
     return [:inactive_ubs] unless ubs.active?
     return [:invalid_schedule_time] if start_time > (Time.zone.now + @max_schedule_time_ahead).at_end_of_day
+    return [:schedule_conditions_unmet] unless patient.can_schedule?
 
-    Appointment.transaction(isolation: :repeatable_read) do
```

Comment:
We should keep the transaction because it's pretty common for timeouts to happen, so the first update may succeed and the others may fail, leaving the database inconsistent. This won't be a performance issue if we add a `SKIP LOCKED`, because the query will return immediately, consequently returning `[:all_slots_taken]` and finishing the transaction.

```sql
UPDATE time_slots
  SET patient_id = $1
  WHERE time_slots = (
    SELECT id
    FROM time_slots
    WHERE patient_id IS NULL
      AND start = $2
      AND ubs_id = $3
    LIMIT 1
    ORDER BY random() -- reduces contention
    FOR UPDATE SKIP LOCKED
  )
  RETURNING id;
```